ft-#162414169 controller for like and dislike articles feature

tosinibrahim96 opened the pull request:
**What does this PR do?**
Creates controller for ArticlesReactions for necessary db operations

**Description of Task to be completed**
A user should be able to like and dislike an article
A user should be able to cancel initial reaction on an article
A user should see all likes for an article they are currently viewing

**How should this be manually tested ?**
Clone the repository and run npm test on command line
After the tests passes, use the command sequelize db:migrate to generate appropriate tables in the database

_To like an article_
**PUT** `localhost:3000/api/v1/articles/reaction/articleId/?reaction=like`

_To dislike an article_
**PUT** `localhost:3000/api/v1/articles/reaction/articleId/?reaction=dislike`

_To get all likes for an article_
**GET** `localhost:3000/api/v1/articles/reaction/articleId/?reaction=like`

_To get all dislikes for an article_
**GET** `localhost:3000/api/v1/articles/reaction/articleId/?reaction=dislike`

_To cancel existing reaction for an article_
**DELETE** `localhost:3000/api/v1/articles/reaction/reactionId`

**Any background context ?**
The reactions controller can only work as expected if the project contains the required
models, migration files and model manager. Make sure you look out for these files.

**What are the relevant pivotal tracker stories ?**
[Users should be able to like and or dislike articles](https://www.pivotaltracker.com/story/show/162414169)

Commit:
[ft-#162414169] implement like and dislike feature for articles

- create article reaction controller for database operations
- write test to verify controller interacts well with the
  modelManager without errors.
- change routes for more clarity
Delivers #162414169

codrex commented on controllers/articlesReactionController.js:
```diff
@@ -0,0 +1,98 @@
+  static async allReactions(req, res) {
+    const message = formatMessage(req.params.reaction);
+    try {
+      const reactions = await getTotalReactions(req.params.articleId, req.params.reaction);
```
`getTotalReactions` This name feels more like we are getting only the total number of reactions. If your interest here is just the total reactions, you should use the count method in sequelizejs.
http://docs.sequelizejs.com/manual/tutorial/models-usage.html#-count-count-the-occurrences-of-elements-in-the-database

tosinibrahim96 replied:
Thanks so much Rex. I just fixed that

codrex commented on controllers/articlesReactionController.js:
```diff
@@ -0,0 +1,98 @@
+    const message = formatMessage(req.params.reaction);
+    try {
+      const reactions = await getTotalReactions(req.params.articleId, req.params.reaction);
+      const { rows, ...count } = reactions;
```
The rest operation here is not needed
```suggestion
      const { rows, count } = reactions;
```

tosinibrahim96 replied:
This is now gone because of the change I made above

codrex commented on middlewares/valueVerifier.js:
```diff
@@ -47,8 +78,12 @@ class valueVerifier {
+      } else if (req.baseUrl === ARTICLE_BASE_URL) {
+        reaction = await getReaction(req.params.reactionId);
+      }
+      if (reaction) {
```
Please, can you explain what is going on here?

tosinibrahim96 replied:
I'm trying to verify if a reaction Id exist before deleting it. The same method is used for both article and comment reactions. I'm trying to check the base URL to ensure I call the right function instead of writing the same function twice

tosinibrahim96 replied:
I finally made use of closure and it's cleaner. Thanks for the help

Commit:
[ft-#162414169] implement like and dislike feature for articles

- create article reaction controller for database operations
- write test to verify controller interacts well with the
  modelManager without errors.
- change routes for more clarity
Delivers #162414169

iMichaelOwolabi reviewed (dismissed):
Looks good

iAfolayan reviewed (commented):
Nice work.

Commit:
[ft-#162414169] fix issues in previous commit

- delete verify reaction file
- Adjust controller to make sure only owner can delete
- remove verifyReaction middleware from route
Delivers #162414169

codrex commented on routes/api/articles.js:
```diff
@@ -297,4 +303,83 @@ router.get(
+ *
+ *
+ */
+router.delete('/reaction/:reactionId', jwtValidator, cancelReaction);
```
The endpoint for this would be `api/v1/comments/some_reactionn_id` this seems to me like you are deleting a comment. Having the endpoint this way `api/v1/comments/reaction/some_reactionn_id`, clearly describes what is going on here.

codrex replied:
A server error is thrown when a reaction id is not valid. Please validate reaction id. Please do the same for every endpoint where ID is required.

tosinibrahim96 replied:
I have made all necessary changes. Thanks

tosinibrahim96 replied:
`api/v1/comments/reaction/some_reactionn_id` for comment reaction
`api/v1/articles/reaction/some_reactionn_id` for article reaction

Commit:
[ft-#162414169] fix issues in previous commit

- Giving a descriptive message when a wrong value is passed as query for reaction.
- The query should accept any case of ‘like’ and ‘dislike’ e.g
   -‘LikE’ and ‘DISLIKE’ should be accepted.
- Getting all reactions for an article should show   both number of likes and dislikes
   - at once instead of using query parameter to fetch them separately.
- A wrong UUID value should show a descriptive message should not be Internal server error
Delivers #162414169
